spec/libraries/duplicator_spec.rb

```diff
@@ -0,0 +1,760 @@
+require 'rails_helper'
+
+RSpec.describe Duplicator, type: :model do
+  context 'when Plain Old Ruby Objects' do
+    class SimpleObject
+      def initialize(id)
+        @id = id
+      end
+
+      def initialize_duplicate(_duplicator)
+      end
+
+      def ==(other)
+        self.class == other.class && state == other.state
+      end
+
+      protected
+
+      def state
+        [@id]
+      end
+    end
+
+    # ComplexObject has children
+    class ComplexObject
+      attr_reader :children
+
+      def initialize(id, children)
+        @id = id
+        @children = children
+      end
+
+      def initialize_duplicate(duplicator)
+        new_children = []
+
+        @children.each do |child|
+          new_child = duplicator.duplicate(child)
+          new_children << new_child unless new_child.nil?
```

Comment:
you can just add all of them to the array, then when you're done call `new_children.compact!`